packages/discovery-provider/src/queries/get_extended_purchase_gate.py

```diff
@@ -0,0 +1,240 @@
+import logging
+from datetime import datetime
+from typing import Dict, List, Optional, TypedDict, Union, cast
+
+from sqlalchemy import and_, func
+from sqlalchemy.orm import aliased
+from sqlalchemy.orm.session import Session
+
+from src.models.playlists.playlist import Playlist
+from src.models.tracks.track import Track
+from src.models.users.user import User
+from src.models.users.user_bank import USDCUserBankAccount
+from src.models.users.user_payout_wallet_history import UserPayoutWalletHistory
+from src.utils.db_session import get_db_read_replica
+from src.utils.session_manager import SessionManager
+
+logger = logging.getLogger(__name__)
+
+
+class GetPurchaseInfoArgs(TypedDict):
+    content_id: int
+    content_type: str
+    content: Union[Track, Playlist]
+
+
+class Split(TypedDict):
+    user_id: int
+    percentage: float
+
+
+class USDCPurchaseCondition(TypedDict):
+    price: int
+    splits: List[Split]
+
+
+class PurchaseGate(TypedDict):
+    usdc_purchase: USDCPurchaseCondition
+
+
+class ExtendedSplit(Split):
+    amount: int
+    payout_wallet: str
+    eth_wallet: str
+
+
+class ExtendedUSDCPurchaseCondition(TypedDict):
+    price: int
+    splits: List[ExtendedSplit]
+
+
+class ExtendedPurchaseGate(TypedDict):
+    usdc_purchase: ExtendedUSDCPurchaseCondition
+
+
+class LegacyUSDCPurchaseCondition(TypedDict):
+    price: int
+    splits: Dict[str, int]
+
+
+class LegacyPurchaseGate(TypedDict):
+    usdc_purchase: LegacyUSDCPurchaseCondition
+
+
+class FollowGate(TypedDict):
+    follow_user_id: int
+
+
+class TipGate(TypedDict):
+    tip_user_id: int
+
+
+class NFTCollection(TypedDict):
```

Comment:
feels like some (all?) of these should be in some models file (or somwhere else more common)

Reply:
You're not wrong... honestly the type system in python is so bad making these types wasn't even worth it...

This PR really made me want to rewrite indexing in Typescript :joy:

I think I'll leave the move of these to any future maintainer's discretion since right now they're not used anywhere else